src/core/layertree/qgslayertreemodellegendnode.cpp

```diff
@@ -648,31 +667,69 @@ void QgsSymbolLegendNode::updateLabel()
+  mLabel = symbolLabel();
 
-  if ( mEmbeddedInParent )
+  if ( showFeatureCount && vl )
   {
-    QString layerName = mLayerNode->name();
-    if ( !mLayerNode->customProperty( QStringLiteral( "legend/title-label" ) ).isNull() )
-      layerName = mLayerNode->customProperty( QStringLiteral( "legend/title-label" ) ).toString();
-
-    mLabel = mUserLabel.isEmpty() ? layerName : mUserLabel;
-    if ( showFeatureCount && vl && vl->featureCount() >= 0 )
-      mLabel += QStringLiteral( " [%1]" ).arg( vl->featureCount() );
+    qlonglong count = mEmbeddedInParent ? vl->featureCount() : vl->featureCount( mItem.ruleKey() ) ;
+    mLabel += QStringLiteral( " [%1]" ).arg( count != -1 ? QLocale().toString( count ) : tr( "N/A" ) );
   }
-  else
+
+  emit dataChanged();
+}
+
+QString QgsSymbolLegendNode::evaluateLabel( const QgsExpressionContext &context, const QString &label )
+{
+  if ( !mLayerNode )
+    return QString();
+
+  QgsVectorLayer *vl = qobject_cast<QgsVectorLayer *>( mLayerNode->layer() );
+
+  if ( vl )
   {
-    mLabel = mUserLabel.isEmpty() ? mItem.label() : mUserLabel;
-    if ( showFeatureCount && vl )
+    QgsExpressionContext contextCopy = QgsExpressionContext( context );
+    QgsExpressionContextScope *symbolScope = createSymbolScope();
+    contextCopy.appendScope( symbolScope );
+    contextCopy.appendScope( vl->createExpressionContextScope() );
+
+    if ( label.isEmpty() )
+    {
+      if ( ! mLayerNode->labelExpression().isEmpty() )
+        mLabel = QgsExpression::replaceExpressionText( "[%" + mLayerNode->labelExpression() + "%]", &contextCopy );
+      else if ( mLabel.contains( "[%" ) )
+      {
+        const QString symLabel = symbolLabel();
+        mLabel = QgsExpression::replaceExpressionText( symLabel, &contextCopy );
+      }
+      return mLabel;
+    }
+    else
     {
-      qlonglong count = vl->featureCount( mItem.ruleKey() );
-      mLabel += QStringLiteral( " [%1]" ).arg( count != -1 ? QLocale().toString( count ) : tr( "N/A" ) );
+      QString eLabel;
+      if ( ! mLayerNode->labelExpression().isEmpty() )
+        eLabel = QgsExpression::replaceExpressionText( label + "[%" + mLayerNode->labelExpression() + "%]", &contextCopy );
+      else if ( label.contains( "[%" ) )
+        eLabel = QgsExpression::replaceExpressionText( label, &contextCopy );
+      return eLabel;
     }
   }
-
-  emit dataChanged();
+  return mLabel;
 }
 
+QgsExpressionContextScope *QgsSymbolLegendNode::createSymbolScope() const
+{
+  QgsVectorLayer *vl = qobject_cast<QgsVectorLayer *>( mLayerNode->layer() );
 
+  QgsExpressionContextScope *scope = new QgsExpressionContextScope( tr( "Symbol scope" ) );
+  scope->addVariable( QgsExpressionContextScope::StaticVariable( QStringLiteral( "symbol_label" ), symbolLabel().remove( "[%" ).remove( "%]" ), true ) );
+  scope->addVariable( QgsExpressionContextScope::StaticVariable( QStringLiteral( "symbol_id" ), mItem.ruleKey(), true ) );
+  if ( vl )
+  {
+    vl->countSymbolFeatures();
```

Comment:
This call will not have any immediate effect for the subsequent `vl->featureCount()` call below. It's probably better to remove it. We can streamline a call to this (with appropriate signals when done and ready to update child labels) via the parent layer node in the future.

Reply:
The goal of the call is to ensure that counting is done at some point if it wasn't done already. It is true that the expression generated will not benefit from it at first but there are functions in place to trigger a refresh once counting is properly done.

For the signal approach what do you suggest?

Reply:
Isn't this keeping a feature counting alive in background always, because every time the counting is finished, all `evaluateLabel()` methods for each legend node will be called, they need to make a new `createSymbolScope()` call and there we have the loop?

Reply:
Yes until a counter is first done and trigger the signal in the layer it will add a new counter. The way around this would be to have a variable in the vectorlayer that prevent adding new counter background task while one is active. 

Is this a type of change that I can do in the vector layer or do you want me to leave it untouched? Otherwise this could probably be done from the data method in the legendmodel but this would be far more annoying to manage.

Reply:
Sidenote: [these two functions are why I cannot rely on the counter directly to get the fids](https://github.com/qgis/QGIS/blob/94f8b65c5615ae91b6e42ab72bd9c8f8ffe5cd2f/src/core/qgsvectorlayer.cpp#L4474)

I will add the counting in the data method of the legendmodel.

Reply:
That doesn't sound safe. The signal is emitted when the counter is *finished* so it's no longer guaranteed to be *active* and therefore using these state indicators doesn't sound like a safe plan.

The trick is to trigger the counting in a method that is only executed once (e.g. initializer method) and not in a data that gets executed every once in a while (like scope generator or data)

Reply:
I think that where I moved the counting should be fine as the data gets called at layer level and will only be trigger for layers to evaluate. 

This mean that a layer will be counted once if it needs to be evaluated instead of each symbol evaluation.

You have to remember that the legend or the feature composition may change. Having too many calls to  `countSymbolFeatures();` is not dangerous because no counting will occur if the stored counts are still valid. 

Otherwise counting only once will probably ensure that either the wrong coung/fids will be provided or that nothing will be return once the stored count/fids or counter is no longer valid.

Reply:
Ok, let's trust the safety guard and cache invalidation in `countSymbolFeatures()` :+1: Can you add a short comment to this call that explains this?

Reply:
Of course!